### PR TITLE
Fixed issue when parmamendIDs was already obtained but NSManagedObject wasn't saved

### PR DESCRIPTION
**Pull request by m1entus**
There can be situation than permamentID was already obtained but context wasn't saved to persistent store so NSManagedObject will have still temporaryID.

1. A in localContext --save--> A in mainContext (not saving to persistent store) --> obtainPermamentID
2. localContext2 <--fetch A -- mainContext  (not saving to persistent store)
3. localContext2 --save--> mainContext
4. MR_inContext:mainContext  called on A from localContext 

MR_inContext will return nil instead of NSManagedObject if he was already obtained .

### CODE REVIEW

**Comment by tonyarnold** [comment]
Thanks @m1entus — is there any chance you can write a test to verify this problem (and fix), so that it doesn't resurface again?

**Comment by m1entus** [comment]
Sure i will do that

**Comment by m1entus** [comment]
Okay, so after writing a test i got core data multithread assertion. Ok so if i am here, could you explain me why it is happening, http://pastebin.com/vz3tVkbv, this is also my test on pastebin. 

If my parent context didn't user MR_obtainPermanentIDsBeforeSaving (which obtain permanent object before save), and will be temporary in defaultContext, i am getting assertion on CoreData. Any ideas why ?

**Comment by m1entus** [comment]
And also i don't see the place where

**Comment by m1entus** [comment]
1. My issue is that from parent contact object is not obtained permamentID.
2. Why you are obtainingPermanentIDsBeforeSaving ?
3. Could you expose MR_obtainPermanentIDsBeforeSaving ?

**Comment by tonyarnold** [comment]
> My issue is that from parent contact object is not obtained permamentID.

I understand, and the test you've written makes that a bit clearer.

> Why you are obtainingPermanentIDsBeforeSaving?

In the circumstance of the method you've updated (`-[NSManagedObject MR_inContext:]`), I believe (it has been a while) temporary `NSManagedObjectID`s do not resolve between contexts.

Given the purpose of `MR_inContext:` (to retrieve a managed object from another context), it seemed wise to obtain the permanent IDs before attempting the lookup.

> Could you expose `MR_obtainPermanentIDsBeforeSaving`?

Yeah, I can — but I'm not sure what it gives you that you can't build yourself using `-[NSManagedObjectContext obtainPermanentIDsForObjects:error]`. What are you intending to use it for?

File: MagicalRecord/Categories/NSManagedObject/NSManagedObject+MagicalRecord.m [hunk 1]
```diff
@@ -287,8 +287,8 @@ - (id) MR_inContext:(NSManagedObjectContext *)otherContext
     
     if ([[self objectID] isTemporaryID])
     {
-        BOOL success = [[self managedObjectContext] obtainPermanentIDsForObjects:@[self] error:&error];
-        if (!success)
+        [[self managedObjectContext] obtainPermanentIDsForObjects:@[self] error:&error];
```
Comment (tonyarnold) [inline MagicalRecord/Categories/NSManagedObject/NSManagedObject+MagicalRecord.m]: I'm a bit worried that if the call to `obtainPermanentIDsForObjects:error:` returns `NO`, this change will ensure that we don't return `nil` here.  That's a significant change in logic.

`+[MagicalRecord handleErrors:]` is safe to call if the `error` parameter is `nil`, so there's no need for it to be wrapped in a check for the existence of the `error` instance.

**Comment by m1entus** [comment]
Current code is fine that's why i closed my PR, i had some issues with my code and temporaryIDs between contexts. 

> Yeah, I can — but I'm not sure what it gives you that you can't build yourself using -[NSManagedObjectContext obtainPermanentIDsForObjects:error]. What are you intending to use it for?

Same issue as here: https://github.com/magicalpanda/MagicalRecord/issues/1084 . I wanted to make main queue context with parent context. Instead of this i had to make:
`+ (instancetype)re_newMainQueueContext {
    NSManagedObjectContext *context = [NSManagedObjectContext MR_newMainQueueContext];
    context.parentContext = [NSManagedObjectContext MR_defaultContext];
    [context re_obtainPermanentIDsBeforeSaving];
    return context;
}`

**Pull request closed without merging**